Added Circle builds and cancel builds functionality

chelsey opened the pull request:
Originally opened here: https://github.com/Shyp/cli/pull/48

chelsey commented:
Seems like I have some import failures - looking

chelsey commented:
Green now

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+    ghUrl, url, status := build.CompareURL, build.BuildURL, build.Status
+
+    // Based on the status of the build, change the color of status print out
+    if inSlice(status, green) {
```
Oh! It's not super obvious, but we should be able to get to this by calling `build.Passed()`, which checks this:

``` go
func (tb TreeBuild) Passed() bool {
    return tb.Status == "success" || tb.Status == "fixed"
}
```

More here: https://godoc.org/github.com/Shyp/go-circle#TreeBuild.Passed

kevinburkeshyp replied:
Maybe we could add similar functions for the other statuses here...

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+
+// CancelBuild cancels a build (as specified by the build number)
+func CancelBuild(org string, project string, buildNum int) string {
+  fmt.Println("\nCanceling build", buildNum, "for", project, "\n")
```
Might be easier to follow as 

``` go
fmt.Printf("\nCanceling build: %d for %s\n\n", buildNum, project)
```

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+// CancelBuild cancels a build (as specified by the build number)
+func CancelBuild(org string, project string, buildNum int) string {
+  fmt.Println("\nCanceling build", buildNum, "for", project, "\n")
+  build, err := circle.CancelBuild(org, project, buildNum)
```
If you're not using build, probably better as:

``` go
_, err := circle.CancelBuild(...)
```

kevinburkeshyp commented on circle.go:
```diff
@@ -150,3 +155,26 @@ func GetBuild(org string, project string, buildNum int) (*CircleBuild, error) {
+		r = body
+	}
+	d := json.NewDecoder(r)
+	var cb CircleBuild
```
Does `cancel` return you the canceled build in the output? I guess that makes sense

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+  // This throws if the branch doesn't exist
+  if err != nil {
+    return err
+  }
```
May be better/more idiomatic as

``` go
if _, err := git.Tip(branch); err != nil {
    return err
}
```

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  fmt.Println("\nFetching recent builds for", branch, "starting with most recent commit\n")
+  // Give CircleCI a little bit of time to start
```
We did this with `wait` to add a little bit of latency between a git push & a Circle build. Not sure we need to sleep here, since we're fetching all recent builds.

kevinburkeshyp commented on builds/build.go:
```diff
@@ -0,0 +1,93 @@
+
+    fmt.Println(url, status, ghUrl)
+
+    sum += i
```
Not sure we're using this var :(

misterMuyiwa commented on builds/build.go:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  fmt.Println("Verify status by running `shyp builds`")
+  return ""
```
might not want this to refer directly to shyp builds since this is a public repo and `shyp builds` is not available

kevinburkeshyp commented:
This is really cool!

kevinburkeshyp commented:
I left some suggestions, but this LGTM either way; I don't want to get in your way, and I'm really excited about this. Let me know if you run into problems pulling this into Shyp CLI.

chelsey commented:
Thanks @kevinburkeshyp and @misterMuyiwa - Really appreciate the feedback as I'm new to this all. Addressed your comments and will merge once the build finishes green 👍
